[Docs] Style guide: General writing and Teleport style suggestions (#31662)
gravitational/teleport: docs/pages/contributing/documentation/style-guide.mdx

- Should be easy to navigate with a browser's search functionality. List all content on the same page.
- Should avoid prose and be expressed with brevity.

## General guidance for engaging content

There are three keys to writing content that engages the audience:

- Use the *second person* point of view to directly address the reader.
- Use an *active voice* and *present tense* whenever possible.
- Use a *conversational tone* that is not too formal or too chummy.

### Point of view

> In most cases, address the reader directly. For tutorials, use either first person 

Reviewer: Should this be "For tutorials and how-to guides"?

plural—we, us, our, ours—or second person point of view. Because tutorials provide a 
more guided approach to a topic, using the first person plural is a more natural and 
commonly-accepted practice than in other types of documentation.

Use the first person point of view sparingly and with intention. When overused, the 
first person narrative can overwhelm the sense of a shared experience and obscure the 
reader's journey.

Don't use “I” or “me” unless it appears in the text of a user interface element.

> Avoid using “we” unless you have a really good rationale for it. For example, if you 

Reviewer: Does this contradict the advice for tutorials two paragraphs above?

are documenting a recommended setting or practice, using “We recommend...” is ambiguous
("we" being who exactly?) and potentially dangerous if something goes wrong. If there
are industry best practices or recommendations, it's okay to indicate that instead of 
using the royal or editorial "we" in documentation.

### Passive constructions

> In spite of the axiom to never use the passive voice, there are situations where a 

Reviewer: Judging by the PRs I review, I'm not sure the reader is necessarily familiar with what passive voice is or that writing guides tend to proscribe it.

passive sentence structure might be appropriate. Don't twist a sentence into knots just 
to avoid a passive construction. Passive voice does have its place, but be wary of using it.

When writing about software, it's often tempting to describe what's happening from the 
code point of view. However, there's almost always a human being with a goal or task to 
complete who is initiating the activity that the software is executing. If you keep this 
human presence in mind, your writing will be more dynamic, easier to follow, and more 

> interesting to read.

Reviewer: I would put this paragraph first in this section and include examples, clarifying that the default would be to avoid passive voice for prose that's more dynamic, easier to follow, etc., but not so much so that you would tie a sentence in knots. We should include examples because this isn't an idea that all docs authors are familiar with.

### Contractions and conversational tone

Contractions are generally acceptable because they give documentation a more natural 
conversational tone—at least for English speakers. Be conscious of when and why you use 
contractions.

> To keep the tone conversational but concise, adhere to the following common-sense guidelines:

Reviewer: ```suggestion
To keep the tone conversational but concise, adhere to the following guidelines:
```
We wouldn't need guidelines if they were common sense.

> - Use common, well-known words whenever possible.

Reviewer: I would include examples here. I'm not sure "common, well-known words" has a self-evident meaning in the context of a specialist tool like Teleport.

> - Don't use flowery language or literary flourish words and phrases like “and so forth”, 

Reviewer: I would say these are more academic than literary. We're not avoiding all literary language if we already say, "Don't twist a sentence into knots".

  “albeit”, “heretofore”, or “thus”.

- Try to be precise in word choice. For example:

  - Don't use “when”—implying eventuality and time—as interchangeable with “if”, which implies 
    the possibility of something happening.
  - Don't use phrases that introduce ambiguity. For example, instead of “When the process completes…” 
    use “After the process completes…”.
  - Think carefully about word choices like using “since” (implying a period of time) instead 
    of “because” (implying cause and result) or using “once” (single occurrence) instead of 

>     “after” (each time).

Reviewer: Does this list belong in a style guide? This sounds like general advice for writing, and I'm not sure that professional writing even consistently avoids the bad examples. I think our style guide should be as short as it can reasonably be to make it easier for authors to follow.

- Avoid using dead language words and phrases even if they are generally accepted as English 

>   words in practice. For example:

Reviewer: What's the reasoning for this? English is filled with load words, so this strikes me as somewhat arbitrary and limiting.

Author: For clarity, expressing what the dead language or abbreviation means in plain English ensures that everyone understands the intended meaning.
For translation (not an issue for Teleport doc perhaps), using plain English can help translators understand the intended meaning and pick an appropriate term.

  - Instead of “i.e.”, use “that is” or rewrite the sentence to make the meaning clear without 
    needing extra qualification.

>   - Instead of “e.g.”, use “for example”.

Reviewer: In parenthetical statements consisting only of examples, I think "e.g.," makes more sense and appears more succinct than "for example."

  - Instead of “via”, use an appropriate English substitute such as “by”, “through”, 
    "from", or “using”.
  - Instead of “etc.”, use “and so on” or revise the content to make the term unnecessary. 
    For example, revise to use *such as* or *like* followed by an example or two.
  - Instead of “caveat”, use an appropriate English substitute such as “notice”, 
    “caution”, or “warning”.

- Avoid adding unnecessary words or phrases. For example:

  - Instead of “In order to”, just use “to”.
  - Instead of “as well as”, just use “and”.
  - Instead of “and then”, just use “then”.
  - Avoid jargon, colloquialisms, and idiomatic phrases.

- Avoid adverbs and subjective statements. For example, don't use words and phrases like 

>   easily, rapidly, simply, quickly.

Reviewer: ```suggestion
  "easily," "rapidly," "simply," and "quickly."
```

  - Experienced developers who ~~truly~~ prefer to skip the tutorial...
  - We can ~~quickly~~ test this code is functioning as expected...

### Headings

All heading levels should use the following conventions:

- Use sentence case. Sentence case means that only the initial word and proper nouns are 
  capitalized.
- Use active, present tense verbs in headings wherever appropriate, especially in the 
  context of tutorials and how-to guides.
- Use noun phrases for conceptual and reference topics whenever possible.

> - Summarize as succinctly as possible the content they contain.

Reviewer: ```suggestion
- Summarize as succinctly as possible the section they introduce.
```
This threw me a little. I was wondering, "How can a heading summarize itself?"

> - Avoid generic headings like overview and introduction, if possible. While generic 

Reviewer: ```suggestion
- Avoid generic headings like "Overview" and "Introduction," if possible. While generic 
```

  headings can be conceptually useful, they add no value to the content or the navigational 
  experience.
- Always contain content. A heading should never be immediately followed by another 
  heading. As a best practice, avoid using headings strictly for navigation.

### Limit heading levels

As a best practice, avoid building an information hierarchy with more than three heading 
levels. Most content can be effectively organized using two internal heading levels, making 
it easier to navigate and scan for relevant topics. 

### Topic titles

> Avoid using gerunds (verbs ending in -ing) in titles and headings. Procedure titles and 

Reviewer: ```suggestion
Avoid using gerunds (verbs ending in "-ing") in titles and headings. Procedure titles and 
```

> headings should answer the question: *What are you trying to do?* For example, if the answer 

Reviewer: ```suggestion
headings should answer the question: *What are you trying to do?*. 

For example, if the answer 
```

cluster. For certain use cases, it may be necessary to adjust our usual voice
for the audience of a specific guide.

### Code, commands, and configuration

- `tsh`, `tctl`, and other core commands should be placed in backticks.

> - All ports or values should be enclosed in backticks, for example, `443`.

Reviewer 2: I'm not a fan of this one. I would prefer we use backticks when we really want a monospaced font (things like code snippets, configuration files, commands, etc). Otherwise we just add extra visual load without good reason.

For example, is:

Teleport listens on port `3080` by default.

really better than

Teleport listens on port 3080 by default.

??
